### ISSUE TITLE
Deprecated declaration warnings in user projects are disabled

### ISSUE THREAD

**Issue by gassmoeller** [dealii/dealii#17812]
It looks like the deal.II deprecation warnings are by default disabled by setting the `-Wno-deprecated-declarations` flag for gcc and clang compilers. And they are not just disabled when compiling deal.II itself, but also for projects that use the `deal_ii_setup_target` macro. I think this may have been an unintentional change in https://github.com/dealii/dealii/pull/14491 because before that PR the flag was set when compiling deal.II itself, but then removed from the `DEAL_II_CXX_FLAGS` when setting up projects that include deal.II (see https://github.com/dealii/dealii/pull/14491/files#diff-afb543fb58675e305cc4ad2a8510f10c024620b036bec9f0dbe877d77ac2b5f9L72). @tamiko: From the conversation in that PR it was not entirely clear to me if this was intentional or not, but it did disable all deprecation warnings in ASPECT, which led to us accumulating a lot of calls to deprecated deal.II functions. I guess my question is: Do we expect user projects to enable deprecation warnings themselves by setting `-Wdeprecated-declarations` or should deal.II export these flags to application projects?

This came up in geodynamics/aspect#6106 and geodynamics/aspect#6115.

**Comment by bangerth** [comment]
Whether intentional or not, it seems wrong. We *want* user projects (and when we compile tutorial programs) to see deprecation warnings. We just don't want them for our own source files and for the test suite.

**Comment by tamiko** [comment]
@bangerth I'll open a pull request that filters `DEAL_II_WARNING_FLAGS` at the correct place again.

The discussion that @masterleinad and I had and that you linked revolved around the fact that both of us agree that we should not export `DEAL_II_WARNING_FLAGS` at all and make it a responsibility of the user project to set warning flags that they need. Because of this we now have the following behavior:
 - the `dealii::dealii_debug` and `dealii::dealii_release` targets will only enforce compiler definitions (such as `DEBUG`) and compiler flags (for example ensuring the correct C++ standard) that are needed to compile and link against the libraries correctly.
 - the target `dealii::dealii` will add warning flags
 - the macro `deal_ii_setup_target()` will also add our warning flags

I will have a look at moving aspect to the low level targets as well.

**Comment by tjhei** [comment]
Just for the record. We need to investigate why test deprecation warnings are not triggering, as can be seen in #17884 for example.

**Comment by bangerth** [comment]
I agree that it makes sense to not export warning flags *in general*. I might disagree on the specific case of deprecation warnings, though. Perhaps we need to do so selectively?
